### ISSUE TITLE
[no-floating-promises] return'ing a promise from within a try block should be illegal

### ISSUE THREAD

**Issue by rix0rrr** [typescript-eslint/typescript-eslint#2557]
The `no-floating-promises` rule is a life-saver in any non-trivial project that uses Promises, because it's so easy to forget to await one.

I just found a bug in our project because of an unawaited Promise that the `no-floating-promises` rule doesn't catch. Probably because the Promise looks used, but actually probably isn't in the way you intend.

The following two definitions of `f` are equivalent, and the rule doesn't throw an error (in the second case, the return Promise of `g()` is considered *handled* and that is true).

```ts
async function g(): Promise<number> { 
  return 3;
}

async function f(): Promise<number> {
  return await g();
}

async function f(): Promise<number> {
  return g();
}
```

However, this is not true anymore if a `try` block is added to `f()`:

```ts
async function f(): Promise<number> {
  try {
    return [await] g();
  } except (e) { 
     // ...
  } finally {
    // ...
  }
}
```

In this case, whether or not the `await` is there makes a difference in whether the execution of `g()` is guarded by the `try` block or not. For the (most likely!) intended behavior, the `await` there is no longer optional -- it should be included to guard exceptions thrown by `g()`.

Proposed behavior: change the rule (or if you consider that too high a risk of breakage, add a flag) to consider the `return` of a Promise from inside a `try` block a violation of the rule as well.

I expect people to argue about whether the current behavior should be classified *intentional* or not. I think *not*, for the following reasons:

* The missing `await` is probably a bug
* It's not obvious whether `return g();` without an `await` is a bug or not. It depends on whether `g()` returns a Promise. The TSC type checker won't help you determine that, so you now need to manually check the types on every function you read in order to know whether an await should be there. That seems like a computer's job to check.
* Just in general: the code is not doing what you would expect from a cursory reading.

If you definitely meant to not have the async execution handled inside the `try` block, there is a simple rewrite to make it obvious that the execution is not guarded (lift the invocation and return outside of the block).

If you are using this rule in the first place, you are probably looking to get some guarantees on your use of Promises, so I think it's worthwhile checking for this.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

Repro already given.

**Expected Result**

I expected `no-floating-promises` to help me avoid mistakes around Promise handling.

**Actual Result**

There is a very sharp edge that requires me to be a 10x Promise Rockstar Ninja, which is exactly what I'm hoping to avoid having to be by using a linter 🙃

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.10.1` |
| `@typescript-eslint/parser`        | `3.10.1` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.12.0` |

**Comment by bradzacher** [comment]
`no-floating-promise` does one thing - it looks for promises that are floating - they have no await and are not "stored" somewhere.

A promise that is returned is not floating. This rule will never handle anything to do with that case.

---

Have you tried a lint rule like `return-await`? 

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/return-await.md

**Comment by rix0rrr** [comment]
A Google search did not turn that up, but that seems more or less like what I'm looking for. Thanks.
